Created Printer classes to support more formats.

hoshir opened the pull request:
Some code are redundant only to support csv format, but wide format will be added soon in the next PR.

Commit:
Created Printer classes to support more formats.

Commit:
A small refactoring for flag validation and completion by introducing Flags interface.

Commit:
Removed a redundant line.

Commit:
style fix

anuraaga commented on cmd/util/array.go:
```diff
@@ -0,0 +1,25 @@
+package util
+
+// Returns true if the given string is in the array.
+func InArray(a []string, s string) bool {
```
Let's call it `Contains`

anuraaga commented on cmd/flag/pass_range_flags.go:
```diff
@@ -0,0 +1,90 @@
+func (f *PassRangeFlags) Complete() error {
+	aosAfter, err := time.Parse(timeFormat, f.flgAOSAfter)
+	if err != nil {
+		aosAfter = time.Now()
```
Can we do the fallback only when the flag isn't provided instead of on a parse error? Let's validate that aosAfter / Before are parsable if specified.

Commit:
Small refactoring.

Commit:
Set aos after/before when it is not provided.

hoshir commented:
Updated

anuraaga commented on cmd/flag/pass_range_flags.go:
```diff
@@ -0,0 +1,110 @@
+	// Maximum value of duration in days.
+	maxDurationInDays uint8 = 31
+	// Time format used to parse "after" and "before" flags.
+	timeFormat = "2006-01-02 15:04:05"
```
Should we make this `2006-01-02T15:04:05`? Otherwise users have to quote the string when specifying it as a flag on the command line since there is a space inside.

anuraaga replied:
Unused

hoshir replied:
This is used in an error message. We need a layout to format AOSAfter and AOSBefore when we want to show it in the message.

anuraaga commented on cmd/flag/pass_range_flags.go:
```diff
@@ -0,0 +1,110 @@
+	}
+
+	// Validate flgAOSAfter when it is provided.
+	if len(f.flgAOSAfter) > 0 {
```
I'd lean towards comparing against empty string since it seems clearer, `f.flgAOSAfter == ""`

anuraaga commented on cmd/groundstation/listPlans.go:
```diff
@@ -41,14 +33,15 @@ var (
+	passRangeFlags := flag.NewPassRangeFlags()
+	outputFormatFlags := flag.NewOutputFormatFlags()
+
+	var flags []flag.Flags
```
`flags := [2]flag.Flags{passRangeFlags, outputFormatFlags}` I think

anuraaga commented on cmd/groundstation/listPlans.go:
```diff
@@ -58,35 +51,37 @@ func NewListPlansCommand() *cobra.Command {
 
-			if flgDuration == 0 || flgDuration > maxDurationInDays {
-				return fmt.Errorf("invalid value of duration: %v. Expected value: 1-%v", flgDuration, maxDurationInDays)
+			for _, f := range flags {
```
Let's define a `ValidateAll` function that does this. Or we can wait and define a wrapper for `Command` instead, which might clean things up more.

Commit:
Created a FlagSet to be cleaner.

hoshir commented:
I created FlagSet that have ValidateAll and AddFlags. I think it make the code cleaner and simple.
Also, I deleted Complete function from Flag and moved logic in it  to validation since it is not very easy to completely separate validate and complete.

Commit:
Style fix.

Commit:
Comment fix

anuraaga commented on cmd/flag/flag_set.go:
```diff
@@ -0,0 +1,49 @@
+// Create a new flag collection.
+func NewFlagSet(args ...Flag) *FlagSet {
+	s := &FlagSet{}
+	for _, f := range args {
```
Think can do `append(s.flags, args...)`

hoshir commented:
Ready for the review again.

anuraaga commented on cmd/util/time.go:
```diff
@@ -0,0 +1,48 @@
+)
+
+// Parse the given time and return time.Time.
+func ParseDateTime(dateStr string) (time.Time, error) {
```
Guess this should return `*time.Time` since we use pointers elsewhere too

anuraaga commented on util/printer/csv_printer.go:
```diff
@@ -0,0 +1,124 @@
+// A Printer which output values as a CSV format.
+type CSVPrinter struct {
+	// Options of CSV format.
+	Options *CSVPrinterOptions
```
Don't think this needs to be pointer

hoshir replied:
I think there isn't standard rule of when to use pointer to struct. The rule I uses is when a struct is small, use struct as a value. When it is a big, I would use pointer. 
In this case, CSVPrinterOptions is small to mid size, but I thought it might have more properties in the future.

anuraaga replied:
I would also say if a struct is only used in one other location, like here where `CSVPrinterOptions` is only used by `CSVPrinter`, it's better not to use a pointer. Go will usually inline everything so there aren't many copies.

Commit:
Resolved comments.

Commit:
Changed CSVPrinterOptions to value passing.

Commit:
Changed time.Time back to value.
